### DISCUSSION TITLE
Async inside visit, or unist-util-visit-async package

### DISCUSSION THREAD

**Discussion by UlisseMini** [syntax-tree/unist#64] [Ideas] [answered]
I'm trying to write a `remarkRunCode` function which replaces javascript code blocks with the `run` tag with their `markdown` export. Attached is my attempt at implementing this, the problem is at the `await visit(tree, "code", async (node, index, parent)` line, `visit` doesn't support async functions so this doesn't work. If I replace `const module = await importInline(node.value)` with `const module = {markdown: "foo *bar* baz"}` it works fine.

```js
function remarkRunCode() {
  return async (tree, file) => {
    visit(tree, "code", async (node, index, parent) => {
      // check for "```js run", the stuff after the space is meta
      if (node.meta === "run") {
        const module = await importInline(node.value);
        const tree = unified().use(remarkParse).parse(module.markdown);
        parent.children.splice(index, 1, ...tree.children);
        console.log(parent.children.at(-1).children);

        // Skip traversing children of 'node' at 'index'
        return [SKIP, index];
      }
    });
  };
}

let cacheBusts = 0;
async function importInline(code) {
  let module;
  try {
    await fs.writeFile(".tmp.js", code);
    module = await import(`./.tmp.js?q=${++cacheBusts}`);
  } finally {
    await fs.remove(".tmp.js");
  }
  return module;
}
```

I fixed it by writing my own visit function (different semantics to the normal `visit` but it works for my usecase)
```js
async function asyncVisit(node, test, callback, index, parent) {
  if (is(node, test)) {
    await callback(node, index, parent);
    // skip children by default, since code blocks don't have children
    // and that's what I'm using this for
  } else if (node.children) {
    for (const i in node.children) {
      await asyncVisit(node.children[i], test, callback, i, node);
    }
  }
}
```
But I feel this should be supported by default, either in `unist-util-visit` or some new package `unist-util-visit-async`.

**Comment by ChristianMurphy** [comment, marked as answer]
Crosslinking a previous version of this question: https://spectrum.chat/unified/syntax-tree/is-there-any-way-to-execute-async-work-when-visiting-a-node-in-unist-util-visit~28177826-628e-44e3-ac3e-0ffb53c195c6

**Reply by UlisseMini** [reply]
Nice! I thought about the `Promise.all` approach but not the way he did it, which is much nicer. Hopefully this question is more searchable since I couldn't find that spectrum thread.
